crates/sdk/src/masp.rs

```diff
@@ -2038,181 +1942,493 @@ impl<U: ShieldedUtils + MaybeSend + MaybeSync> ShieldedContext<U> {
+        // identical values, so overwriting should not be problematic.
+        self.nf_map.extend(new_ctx.nf_map);
+        self.note_map.extend(new_ctx.note_map);
+        self.memo_map.extend(new_ctx.memo_map);
+        self.div_map.extend(new_ctx.div_map);
+        self.witness_map.extend(new_ctx.witness_map);
+        self.spents.extend(new_ctx.spents);
+        self.asset_types.extend(new_ctx.asset_types);
+        self.vk_map.extend(new_ctx.vk_map);
+        // if we successfully finished scanning, we can clear
+        // out viewing keys from our unscanned data and add them to the
+        // pos map.
+        if !self.interrupted() {
+            self.pos_map.extend(new_ctx.pos_map);
+            self.unscanned.vks.clear();
+        }
+        // The deltas are the exception because different keys can reveal
+        // different parts of the same transaction. Hence each delta needs to be
+        // merged separately.
+        for (height, (ep, ntfer_delta, ntx_delta)) in new_ctx.delta_map {
+            let (_ep, tfer_delta, tx_delta) = self
+                .delta_map
+                .entry(height)
+                .or_insert((ep, TransferDelta::new(), TransactionDelta::new()));
+            tfer_delta.extend(ntfer_delta);
+            tx_delta.extend(ntx_delta);
+        }
+    }
+
+    /// Fetch the current state of the multi-asset shielded pool into a
+    /// ShieldedContext
+    pub async fn fetch<C: Client + Sync>(
+        &mut self,
+        client: &C,
+        io: &impl Io,
+        last_query_height: Option<BlockHeight>,
+        sks: &[ExtendedSpendingKey],
+        fvks: &[ViewingKey],
+    ) -> Result<(), Error> {
+        // First determine which of the keys requested to be fetched are new.
+        // Necessary because old transactions will need to be scanned for new
+        // keys.
+        let mut unknown_keys = Vec::new();
+        for esk in sks {
+            let vk = to_viewing_key(esk).vk;
+            if !self.pos_map.contains_key(&vk) {
+                unknown_keys.push(vk);
+            }
+        }
+        for vk in fvks {
+            if !self.pos_map.contains_key(vk) {
+                unknown_keys.push(*vk);
+            }
+        }
+
+        // If unknown keys are being used, we need to scan older transactions
+        // for any unspent notes
+        if !unknown_keys.is_empty() || !self.unscanned.vks.is_empty() {
+            display_line!(
+                io,
+                "Scanning historical notes to synchronize new viewing key(s)."
+            );
+            self.unscanned.vks.extend(unknown_keys.clone());
+            // Load all transactions accepted until this point
+            let fetched = self
+                .fetch_shielded_transfers(
+                    client,
+                    io,
+                    self.latest_unscanned(),
+                    last_query_height,
+                )
+                .await?;
+            self.unscanned.txs.extend(fetched);
```

Comment:
Before this line, is `self.unscanned.txs` supposed to contain all transactions up till `self.latest_unscanned()`? After this line, is `self.unscanned.txs` supposed to contain all transactions up till `last_query_height`?

Reply:
`self.latest_unscanned()` is supposed to tell us the most recent block height in the local cache. However, there is a bug here. We need to fetch from zero up to `self.last_fetched` since we are trying to sync up new keys with existing keys.